crates/astria-account-monitor/src/account_monitor.rs

```diff
@@ -0,0 +1,296 @@
+        else {
+            return Ok(());
+        };
+
+        Ok(())
+    }
+}
+
+async fn run_loop(
+    metrics: &'static Metrics,
+    client: &sequencer_client::HttpClient,
+    accounts: &SequencerAccountsToMonitor,
+    asset: &Asset,
+    pull_interval: Duration,
+) {
+    let mut poll_timer = interval(pull_interval);
+    poll_timer.set_missed_tick_behavior(tokio::time::MissedTickBehavior::Skip);
+
+    loop {
+        poll_timer.tick().await;
+
+        fetch_all_info(metrics, client, accounts, asset);
+    }
+}
+
+fn fetch_all_info(
+    metrics: &'static Metrics,
+    client: &sequencer_client::HttpClient,
+    accounts: &SequencerAccountsToMonitor,
+    asset: &Asset,
+) {
+    for account in accounts {
+        let client = client.clone();
+        let asset = asset.clone();
+        let address = account.address;
+        let _handle =
+            tokio::spawn(async move { fetch_account_info(metrics, &client, address, asset).await });
+    }
+}
+
+/// Note: The return value of this function exists only to be emitted as part of instrumentation.
+#[instrument(skip_all, fields(%address, %asset), ret(Display))]
+async fn fetch_account_info(
+    metrics: &'static Metrics,
+    client: &sequencer_client::HttpClient,
+    address: Address,
+    asset: Asset,
+) -> AccountInfo {
+    metrics.increment_nonce_fetch_count();
+    metrics.increment_balance_fetch_count();
+    let (nonce, balance) = tokio::join!(
+        timeout(
+            Duration::from_millis(1000),
+            get_latest_nonce(client, address)
+        ),
+        timeout(
+            Duration::from_millis(1000),
+            get_latest_balance(client, address, asset)
+        )
+    );
+
+    let account_nonce = match nonce {
+        Ok(Ok(nonce)) => {
+            metrics.set_account_nonce(&address.into(), nonce);
+            QueryResponse::Value(nonce)
+        }
+        Ok(Err(err)) => {
+            warn!(%err, "failed to get nonce");
+            metrics.increment_nonce_fetch_failure_count();
+            QueryResponse::Error
+        }
+        Err(err) => {
+            warn!(%err, "nonce query timed out");
+            metrics.increment_nonce_fetch_failure_count();
+            QueryResponse::Timeout
+        }
+    };
+
+    let account_balance = match balance {
+        Ok(Ok(balance)) => {
+            metrics.set_account_balance(&address.into(), balance.balance);
+            QueryResponse::Value(balance.balance)
+        }
+        Ok(Err(err)) => {
+            warn!(%err, "failed to get balance");
+            metrics.increment_balance_fetch_failure_count();
+            QueryResponse::Error
+        }
+        Err(err) => {
+            warn!(%err, "balance query timed out");
+            metrics.increment_balance_fetch_failure_count();
+            QueryResponse::Timeout
+        }
+    };
+
+    AccountInfo {
+        nonce: account_nonce,
+        balance: account_balance,
+    }
+}
+
+#[derive(Debug, Clone)]
+pub enum QueryResponse<T> {
```

Comment:
I like this ^.^ But don't make this `pub` (we try to keep the services very locked down); if you need to share it across the crate, `pub(crate)`.
```suggestion
enum QueryResponse<T> {
```